examples/kotlin/src/main/kotlin/com/example/booktest/postgresql/QueriesImpl.kt

```diff
@@ -0,0 +1,293 @@
+      val ret = mutableListOf<Book>()
+      while (results.next()) {
+          ret.add(Book(
+      results.getInt(1),
+      results.getInt(2),
+      results.getString(3),
+      BookType.lookup(results.getString(4))!!,
+      results.getString(5),
+      results.getInt(6),
+      results.getObject(7, OffsetDateTime::class.java),
+      (results.getArray(8).array as Array<String>).toList()
+    ))
+      }
+      ret
+    }
+  }
+
+  @Throws(SQLException::class)
+  fun createAuthor(name: String): Author {
+    val stmt = conn.prepareStatement(createAuthor)
+    stmt.setString(1, name)
+
+    return stmt.executeQuery().use { results ->
+      if (!results.next()) {
+        throw SQLException("no rows in result set")
+      }
+      val ret = Author(
+      results.getInt(1),
+      results.getString(2)
+    )
+      if (results.next()) {
+          throw SQLException("expected one row in result set, but got many")
+      }
+      ret
+    }
+  }
+
+  @Throws(SQLException::class)
+  fun createBook(arg: CreateBookParams): Book {
+    val stmt = conn.prepareStatement(createBook)
+    stmt.setInt(1, arg.authorId)
+    stmt.setString(2, arg.isbn)
+    stmt.setObject(3, arg.booktype.value, Types.OTHER)
+    stmt.setString(4, arg.title)
+    stmt.setInt(5, arg.year)
+    stmt.setObject(6, arg.available)
+    stmt.setArray(7, conn.createArrayOf("pg_catalog.varchar", arg.tags.toTypedArray()))
+
+    return stmt.executeQuery().use { results ->
+      if (!results.next()) {
+        throw SQLException("no rows in result set")
+      }
+      val ret = Book(
+      results.getInt(1),
+      results.getInt(2),
+      results.getString(3),
+      BookType.lookup(results.getString(4))!!,
+      results.getString(5),
+      results.getInt(6),
+      results.getObject(7, OffsetDateTime::class.java),
+      (results.getArray(8).array as Array<String>).toList()
+    )
+      if (results.next()) {
+          throw SQLException("expected one row in result set, but got many")
+      }
+      ret
+    }
+  }
+
+  @Throws(SQLException::class)
+  fun deleteBook(bookId: Int) {
+    val stmt = conn.prepareStatement(deleteBook)
+    stmt.setInt(1, bookId)
+
+    stmt.execute()
+    stmt.close()
+  }
+
+  @Throws(SQLException::class)
+  fun getAuthor(authorId: Int): Author {
+    val stmt = conn.prepareStatement(getAuthor)
+    stmt.setInt(1, authorId)
+
+    return stmt.executeQuery().use { results ->
+      if (!results.next()) {
+        throw SQLException("no rows in result set")
+      }
+      val ret = Author(
+      results.getInt(1),
+      results.getString(2)
+    )
+      if (results.next()) {
+          throw SQLException("expected one row in result set, but got many")
+      }
+      ret
+    }
+  }
+
+  @Throws(SQLException::class)
+  fun getBook(bookId: Int): Book {
+    val stmt = conn.prepareStatement(getBook)
+    stmt.setInt(1, bookId)
+
+    return stmt.executeQuery().use { results ->
+      if (!results.next()) {
```

Comment:
I see this pattern a lot. could probably add some kind of extension to `ResultSet`

```
fun ResultSet.uniqueResultNext() {
  if (!results.next()) {
    throws SQLException("no results");
  }
  if (results.next()) {
    throw SQLException("too many");
  }
  results.previous()
}

jdbc is gross
```

Reply:
Agree, but I want to defer needing a library to use this code.